wasm/src/main/java/com/dylibso/chicory/wasm/Parser.java

```diff
@@ -931,9 +934,25 @@ public static long readFloat32(ByteBuffer buffer) {
      * @return
      */
     public static String readName(ByteBuffer buffer) {
+        return readName(buffer, false);
+    }
+
+    public static String readName(ByteBuffer buffer, boolean allow) {
         var length = (int) readVarUInt32(buffer);
         byte[] bytes = new byte[length];
-        buffer.get(bytes);
-        return new String(bytes, StandardCharsets.UTF_8);
+        try {
+            buffer.get(bytes);
+        } catch (BufferUnderflowException e) {
+            throw new MalformedException("length out of bounds");
+        }
+        var name = new String(bytes, StandardCharsets.UTF_8);
+        if (!isValidIdentifier(name) && !allow) {
+            throw new MalformedException("malformed UTF-8 encoding");
+        }
+        return name;
+    }
+
+    private static boolean isValidIdentifier(String string) {
+        return string.chars().allMatch(ch -> ch < 0x80 || Character.isUnicodeIdentifierPart(ch));
```

Comment:
The specification says only "[Names](https://webassembly.github.io/spec/core/syntax/values.html#syntax-name) are encoded as a [vector](https://webassembly.github.io/spec/core/binary/conventions.html#binary-vec) of bytes containing the [Unicode](https://www.unicode.org/versions/latest/) (Section 3.9) UTF-8 encoding of the name’s character sequence." Note that they do not specify any characters as being valid or invalid; therefore, it's not correct to restrict the valid character space of names in any way AFAICT.

Reply:
I see your point and indeed it's very likely that this is not 100% compliant with the spec and that we're leaving out some valid characters. That said @andreaTP and me have already spent a quite relevant amount of time on this and we couldn't find a better solution. I believe that there isn't much value added in spending more time on it at this point. We are aware that we are probably leaving out some valid characters. Nevertheless my suggestion is to merge this pull request as it is, move forward, and come back to this only when we will have a test evidence that this validation isn't working as it should in all cases.

Reply:
My point is that name validation is completely unnecessary, because all Unicode strings are allowed by spec (even the empty name). Adding additional restrictions changes the implementation from being strictly correct to being strictly incorrect. It's not a matter of degrees. Furthermore as I stated above, #249 removes this code from `Parser` in any event.

Reply:
> My point is that name validation is completely unnecessary, because all Unicode strings are allowed by spec.

This is untrue. If I try to remove this validation I have a total of 528 test failures in `SpecV1Utf8ImportModuleTest`, `SpecV1Utf8ImportFieldTest` and `SpecV1Utf8CustomSectionIdTest`, all expecting a `MalformedException` to be throw but without the validation isn't.

Reply:
According to [https://webassembly.github.io/spec/core/syntax/values.html#names](the Names section of the spec), the name codepoints are expected to be within the range of Unicode; perhaps there is a code point outside of that range? Perhaps it's not the character but the encoding? According to [the binary format spec for Names](https://webassembly.github.io/spec/core/binary/values.html#names) it seems to imply that the shortest encoding is required, so maybe the expected failures are related to using longer-than-minimal encoding?

It'd be really good to understand why the test fail rather than patch it over IMO.

Reply:
If you're in doubt try this pull request giving a run to one of the tests I mentioned. For instance if you execute `SpecV1Utf8ImportModuleTest.test0` you will see that at some point the import section reads a name with a single character having codepoint 65533 that is invalid and correctly recognized as such by that method.

Reply:
Ah so I think my suspicion is correct. The code point isn't invalid, its encoding is.

In UTF-8, each character can be encoded in potentially many ways. The leading bits of each UTF-8 character determine how many bits there are in the final character value. However, it is possible to encode a Unicode code point into a UTF-8 sequence that is longer than necessary. It is also possible to provide byte sequences that are invalid. That appears to be what is tested by this test. The JDK will attempt to accept these values in various ways, so in order to property validate, we need a completely different approach to UTF-8 decoding which validates each byte more strictly than what the JDK is doing.

Reply:
Actually @andreaTP already attempted to do [here](https://github.com/mariofusco/chicory/pull/1/files#diff-d5bf04fbcefd84882667b907d29a2b88ceaa9ff6fb176408f3b5dc6f5f24883f) something similar to what (I believe) you're suggesting, but we didn't manage to make it 100% work and anyway it seemed too much complicated at least for our current needs. My suggestion is to avoid wasting more time on this for now (I guess we have more important things to do), merge what we have and eventually come back to this when and if necessary.